src/coreclr/jit/helperexpansion.cpp

```diff
@@ -595,7 +595,10 @@ bool Compiler::fgExpandThreadLocalAccessForCallNativeAOT(BasicBlock** pBlock, St
 
         CORINFO_CONST_LOOKUP tlsIndexObject = threadStaticInfo.tlsIndexObject;
 
+        // We want tls_index to get relocated and hence create IconHandle node,
+        // but it does not have GC ref, so reset the type to TYP_I_IMPL instead of TYP_REF
         GenTree* dllRef = gtNewIconHandleNode((size_t)tlsIndexObject.handle, GTF_ICON_OBJ_HDL);
+        dllRef->gtType  = TYP_I_IMPL;
```

Comment:
What is this the value of? `GTF_ICON_OBJ_HDL` is for non-GC objects. The JIT can call various object related JIT-EE APIs on this value, so this should probably use a different handle type and not have this retyping. If you just want a generic handle type we have `GTF_ICON_CONST_PTR` for pointers to generic immutable data and `GTF_ICON_GLOBAL_PTR` for pointers to generic mutable data.

Reply:
This is the value of `_tls_index`. If I keep it as `IconHandle()`, it's type is set to `TYP_REF` and we end up adding gc tracking for this, but since it is a non-gc native pointer, we don't want to do that. Also I wanted to make this as handle, so it can be relocatable. I will see if I can use a different handle. We set the type to `TYP_REF` only for `GTF_ICON_OBJ_HDL` and `TYP_I_IMPL` otherwise, so definitely a different handle type makes sense.